## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [23]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's read the dataset into a <em>pandas</em> dataframe.

In [24]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [25]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [26]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [27]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [28]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [29]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [30]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [31]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [32]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Import Keras

#### Let's go ahead and import the Keras library

In [33]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [34]:
from keras.models import Sequential
from keras.layers import Dense

In [35]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [36]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [37]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [38]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [39]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
 - 0s - loss: 1622.0356
Epoch 2/50
 - 0s - loss: 1606.6490
Epoch 3/50
 - 0s - loss: 1590.9841
Epoch 4/50
 - 0s - loss: 1575.7692
Epoch 5/50
 - 0s - loss: 1560.0975
Epoch 6/50
 - 0s - loss: 1544.2897
Epoch 7/50
 - 0s - loss: 1527.9925
Epoch 8/50
 - 0s - loss: 1511.2347
Epoch 9/50
 - 0s - loss: 1493.5820
Epoch 10/50
 - 0s - loss: 1475.2346
Epoch 11/50
 - 0s - loss: 1456.0975
Epoch 12/50
 - 0s - loss: 1435.6287
Epoch 13/50
 - 0s - loss: 1414.5105
Epoch 14/50
 - 0s - loss: 1391.8601
Epoch 15/50
 - 0s - loss: 1368.5500
Epoch 16/50
 - 0s - loss: 1344.6083
Epoch 17/50
 - 0s - loss: 1318.8334
Epoch 18/50
 - 0s - loss: 1293.0718
Epoch 19/50
 - 0s - loss: 1265.9818
Epoch 20/50
 - 0s - loss: 1238.7713
Epoch 21/50
 - 0s - loss: 1210.4307
Epoch 22/50
 - 0s - loss: 1182.4510
Epoch 23/50
 - 0s - loss: 1152.9748
Epoch 24/50
 - 0s - loss: 1123.8830
Epoch 25/50
 - 0s - loss: 1093.8568
Epoch 26/50
 - 0s - loss: 1064.4033
Epoch 27/50
 - 0s - loss: 1034.0760
Epoch 28/50
 - 0s - loss: 1004.3933
E

Next we need to evaluate the model on the test data.

In [40]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 339us/step


458.59951683464175

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [41]:
from sklearn.metrics import mean_squared_error

In [42]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

458.59952563373184 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [43]:
total_mean_squared_errors = 50
epochs = 100
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 107.08938850476903
MSE 2: 92.64591723198258
MSE 3: 57.06216962746432
MSE 4: 43.271584013904956
MSE 5: 40.587367406554975
MSE 6: 44.256707077273276
MSE 7: 42.816397978649945
MSE 8: 32.33769681615737
MSE 9: 36.601891884911794
MSE 10: 37.18744454491871
MSE 11: 36.07667916800983
MSE 12: 30.756194636273925
MSE 13: 40.926298110616244
MSE 14: 39.59313725344957
MSE 15: 35.17515849758506
MSE 16: 28.595407819284976
MSE 17: 35.3407456604794
MSE 18: 34.92029393760903
MSE 19: 34.75231773722133
MSE 20: 34.24421805471278
MSE 21: 30.850252009518325
MSE 22: 34.404132151680855
MSE 23: 29.151783680838676
MSE 24: 30.96719287514301
MSE 25: 35.81963603936353
MSE 26: 34.50429911443717
MSE 27: 30.99149681757955
MSE 28: 30.517114935569392
MSE 29: 34.57569354406067
MSE 30: 34.23537195693328
MSE 31: 30.882272016654895
MSE 32: 31.330439083012948
MSE 33: 31.851783937620887
MSE 34: 33.75386659690091
MSE 35: 32.93508958044947
MSE 36: 38.767993760340424
MSE 37: 29.208827219348894
MSE 38: 35.334473236479035
MSE

## Answer

|  | Peer-graded Assignment  | Mean MSE| Standard Deviation | Answer |
| ------------------ | ------- | -------- | ------- |
| A               | Build a baseline model       | 47.96254        | 9.48824      |  |
| B      | Normalize the data        | 43.7538        | 24.77679      | Step B improved the mean of the MSE compared to Step A |
| C                |  Increate the number of epochs (100)       | 37.5862       | 13.67702     | Step C improved the mean of the MSE compared to Step B |
| D |  Increase the number of hidden layers (3)       | ?        | ?    |  |